sdk/python/starwhale/data_store.py

```diff
@@ -0,0 +1,703 @@
+        return value
+
+    def deserialize(self, value: Any) -> Any:
+        return value
+
+    def __str__(self) -> str:
+        if self.name == "int" or self.name == "float":
+            return f"{self.name}{self.nbits}"
+        else:
+            return self.name
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+
+class LinkType(Type):
+    def __init__(self):
+        super().__init__("link", pa.string(), 32, "")
+
+    def serialize(self, value: Any) -> Any:
+        if value is None:
+            return None
+        assert isinstance(value, Link)
+        return str(value)
+
+    def deserialize(self, value: Any) -> Any:
+        if value is None:
+            return None
+        d = json.loads(value)
+        return Link(d.get("uri", None), d.get("display_text", None), d.get("mime_type", None))
+
+
+class Link:
+    def __init__(self, uri: str, display_text: Optional[str] = None, mime_type: Optional[str] = None):
+        self.uri = uri
+        self.display_text = display_text
+        self.mime_type = mime_type
+
+    def __str__(self) -> str:
+        return json.dumps({"uri": self.uri, "display_text": self.display_text, "mime_type": self.mime_type})
+
+    def __eq__(self, other) -> bool:
+        return (
+            isinstance(other, Link)
+            and self.uri == other.uri
+            and self.display_text == other.display_text
+            and self.mime_type == other.mime_type
+        )
+
+
+NONE = Type("none", None, 1, None)
+INT8 = Type("int", pa.int8(), 8, 0)
+INT16 = Type("int", pa.int16(), 16, 0)
+INT32 = Type("int", pa.int32(), 32, 0)
+INT64 = Type("int", pa.int64(), 64, 0)
+FLOAT16 = Type("float", pa.float16(), 16, 0.0)
+FLOAT32 = Type("float", pa.float32(), 32, 0.0)
+FLOAT64 = Type("float", pa.float64(), 64, 0.0)
+BOOL = Type("bool", pa.bool_(), 1, 0)
+STRING = Type("str", pa.string(), 32, "")
+BYTES = Type("bytes", pa.binary(), 32, b"")
+LINK = LinkType()
+
+_TYPE_DICT: dict[Type, Type] = {
+    type(None): NONE,
+    np.byte: INT8,
+    np.int8: INT8,
+    np.int16: INT16,
+    np.int32: INT32,
+    np.int64: INT64,
+    int: INT64,
+    np.float16: FLOAT16,
+    np.float32: FLOAT32,
+    np.float_: FLOAT64,
+    float: FLOAT64,
+    np.bool_: BOOL,
+    bool: BOOL,
+    str: STRING,
+    bytes: BYTES,
+    Link: LINK,
+}
+
+_TYPE_NAME_DICT = {str(v): v for k, v in _TYPE_DICT.items()}
+
+
+def _get_type(obj: Any) -> Type:
+    return _TYPE_DICT.get(type(obj), None)
+
+
+class ColumnSchema:
+    def __init__(self, name: str, type: Type):
+        self.name = name
+        self.type = type
+
+    def __eq__(self, other) -> bool:
+        return isinstance(other, ColumnSchema) and self.name == other.name and self.type is other.type
+
+
+class TableSchema:
+    def __init__(self, key_column: str, columns: list[ColumnSchema]):
+        self.key_column = key_column
+        self.columns = {col.name: col for col in columns}
+
+    def copy(self) -> "TableSchema":
```

Comment:
when write `from __future__ import annotations` at the first line,  `"TableSchema"` can be replaced with `TableSchema`. It is friendly to IDE.

Reply:
I prefer the old way so that it can work in a low-version Python environment. IDE can handle this old-fashion typing annotation well.